### ISSUE TITLE
Spam Assassin Server: IPv6 related error in Perl when invoking sa-update

### ISSUE THREAD

**Issue by mikael-andersson** [NixOS/nixpkgs#20450]
## Issue description

When running sa-update, I get the following error:

> Use of uninitialized value in subroutine entry at /nix/store/mj2i6iayh4n39b08v8r5b41fh1b43as4-perl-5.22.2/lib/perl5/5.22.2/x86_64-linux-thread-multi/Socket.pm line 844.
> Bad arg length Socket::pack_sockaddr_in6, length is 0, should be 16 at /nix/store/mj2i6iayh4n39b08v8r5b41fh1b43as4-perl-5.22.2/lib/perl5/5.22.2/x86_64-linux-thread-multi/Socket.pm line 844.
> 

The version of Spam Assassin is 3.4.1 and it seems to be the latest stable version.  My guess is that the Perl version is a bit too old. What I can see from spamassassin.nix, no perl package and version is specified. Is that really right?

### Steps to reproduce
Activate Spam Assassin Server in configuration.nix with:
services.spamassassin.enable = true;

Run 'sa-update'.

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
16.09.1026.5ddaca1 (Flounder)
* Nix version: (run `nix-env --version`)
nix-env (Nix) 1.11.4
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
"16.09.1026.5ddaca1"

**Comment by mikael-andersson** [comment]
I tried to lookup the required version of Perl and that does not seem to be the problem since it says:

"Perl 5.8.1 or a later version is required."
(Above is an excerpt from: https://github.com/apache/spamassassin/blob/3.4/INSTALL)

Instead, it seems like the latest Perl version in nixpkgs for release 16.09 is used:
https://github.com/NixOS/nixpkgs/blob/release-16.09/pkgs/development/interpreters/perl/default.nix

I found a similar problem reported here: https://rt.cpan.org/Public/Bug/Display.html?id=65599

According to the ticket above, the function (in my it is pack_sockaddr_in6 and not pack_sockaddr_in) actually it is is called with the wrong type.

**Comment by mikael-andersson** [comment]
I have looked a bit more in to this and found this:

```
[root@nixos:/home/mikael]# sa-update -D --updatedir /tmp/updates
Nov 24 18:55:10.184 [15815] dbg: logger: adding facilities: all
Nov 24 18:55:10.184 [15815] dbg: logger: logging level is DBG
Nov 24 18:55:10.184 [15815] dbg: generic: SpamAssassin version 3.4.1
Nov 24 18:55:10.184 [15815] dbg: generic: Perl 5.022002, PREFIX=/nix/store/xnm7i3zgjjac7f3aw1sjqn2830hrkwpv-perl-SpamAssassin-3.4.1, DEF_RULES_DIR=/nix/store/xnm7i3zgjjac7f3aw1sjqn2830hrkwpv-perl-SpamAssassin-3.4.1/share/spamassassin, LOCAL_RULES_DIR=/etc/spamassassin, LOCAL_STATE_DIR=/var/lib/spamassassin
Nov 24 18:55:10.184 [15815] dbg: config: timing enabled
Nov 24 18:55:10.187 [15815] dbg: config: score set 0 chosen.
Nov 24 18:55:10.194 [15815] dbg: generic: sa-update version svn1652181
Nov 24 18:55:10.194 [15815] dbg: generic: using update directory: /tmp/updates
Nov 24 18:55:10.252 [15815] dbg: diag: perl platform: 5.022002 linux
...
```

What makes me a bit confused is that it says Perl 5.022002 here. Any ideas why?

**Comment by mikael-andersson** [comment]
After having seen that one can run sa-update with only IPv4 support, I got an idea, since I got a 'Bad arg length Socket::pack_sockaddr_in6', I invoked `sa-update -4`. This made sa-update to complain about:

config: no configuration text or files found! do you need to run 'sa-update'?
Timeout::_run: check: no loaded plugin implements 'check_main': cannot scan!
Check that the necessary '.pre' files are in the config directory.
At a minimum, v320.pre loads the Check plugin which is required.

This looked much more promising. In spamassassin.nix, I found `CONFDIR=/etc/spamassassin`. I verified that this directory did not contain any configuration and copied v320.pre from $(dirname $(readlink /run/current-system/sw/bin/spamd))/../share/spamassassin/

Now sa-update -4 seems to update successfully.

To me it seems to be a bug when running sa-update with IPv6 support.

I have not debugged the Perl more so I really do not know why `Socket::pack_sockaddr_in6` is 0 length.

I do not have much knowledge about IPv6 but it seems I have been assigned an IPv6 address on my active network interface, so the problem is probably not that an IPv6 address is missing. Although an IPv6 address would be missing, I do not think one should get such a Perl error.

**Comment by Profpatsch** [comment]
(triage) can you still reproduce?

**Comment by plutotulp** [comment]
I guess this is maybe related: I tried running spamd, but whenever I tried connecting with spamc, these messages popped up in the journal:

```
spamd: error: Bad arg length for Socket::unpack_sockaddr_in, length is 28, should be 16 at /
nix/store/4jgv6q38ac1wsigy0gnc1x9nsfjqmby9-perl-5.30.0/lib/perl5/5.30.0/x86_64-linux-thread-multi/Socket.pm line 848
, continuing at /nix/store/rrwj9qzy96xp4z7sk58yv6w88hr0p13r-perl5.30.0-SpamAssassin-3.4.2/b
in/.spamd-wrapped line 1419.
```
and the action fails.

Fixed it by adding the Socket6 module to the buildInputs of spamassassin.

**Comment by alyssais** [comment]
Should have been fixed by #76467.
